### PR TITLE
Experimental features

### PR DESCRIPTION
**Pull request by Hosch250**

### CODE REVIEW

File: RetailCoder.VBE/Navigation/CodeExplorer/CodeExplorerViewModel.cs [hunk 1]
```diff
@@ -558,7 +558,7 @@ private void ExecuteRemoveComand(object param)
 
         public Visibility ExportAllVisibility => CanExecuteExportAllCommand ? Visibility.Visible : Visibility.Collapsed;
 
-        public bool IsSourceControlEnabled => _generalSettings.IsSourceControlEnabled;
+        public bool EnableExperimentalFeatures => _generalSettings.EnableExperimentalFeatures;
```
Comment (retailcoder) [inline RetailCoder.VBE/Navigation/CodeExplorer/CodeExplorerViewModel.cs]: hmm, I think we should list all "experimental features," and allow enabling/disabling them selectively. this way you can have experimental source control without the experimental code panes... or vice-versa.

Comment (Hosch250) [reply RetailCoder.VBE/Navigation/CodeExplorer/CodeExplorerViewModel.cs]: That'd be nice, but pretty impossible without adding an attribute for each feature, and then it gets where we might as well leave it the way it currently is.

Comment (retailcoder) [reply RetailCoder.VBE/Navigation/CodeExplorer/CodeExplorerViewModel.cs]: I'm not following. We could put an `[Experimental("Feature_ResourceKey")]` attribute on top of a class, and the settings viewmodel would reflect and locate all the classes with that attribute, get the corresponding resource string, and stick that into the settings tab for experimental features, no?

File: RetailCoder.VBE/Root/RubberduckIoCInstaller.cs [hunk 2]
```diff
@@ -198,11 +220,21 @@ private static void ApplyDefaultInterfaceConvention(IWindsorContainer container,
...
-                container.Register(Types.FromAssembly(assembly)
+                var assemblyTypes = assembly.DefinedTypes
+                    .Where(w =>
+                    {
+                        var attribute = w.CustomAttributes.FirstOrDefault(f => f.AttributeType == typeof(ExperimentalAttribute));
+                        var ctorArg = attribute?.ConstructorArguments.Any() == true ? (string)attribute.ConstructorArguments.First().Value : string.Empty;
+
+                        return attribute == null ||
+                               _initialSettings.EnableExperimentalFeatures.Any(a => a.Key == ctorArg && a.IsEnabled);
+                    });
```
Comment (Vogel612) [inline RetailCoder.VBE/Root/RubberduckIoCInstaller.cs]: Would it be useful to extract the construction of "available" assemblyTypes into a method?

File: RetailCoder.VBE/Root/RubberduckIoCInstaller.cs [hunk 3]
```diff
@@ -219,34 +251,64 @@ private static void RegisterSpecialFactories(IWindsorContainer container)
...
-                container.Register(Classes.FromAssembly(assembly)
+                var assemblyTypes = assembly.DefinedTypes
+                    .Where(w =>
+                    {
+                        var attribute = w.CustomAttributes.FirstOrDefault(f => f.AttributeType == typeof(ExperimentalAttribute));
+                        var ctorArg = attribute?.ConstructorArguments.Any() == true ? (string)attribute.ConstructorArguments.First().Value : string.Empty;
+
+                        return attribute == null ||
+                               _initialSettings.EnableExperimentalFeatures.Any(a => a.Key == ctorArg && a.IsEnabled);
+                    });
```
Comment (Vogel612) [inline RetailCoder.VBE/Root/RubberduckIoCInstaller.cs]: This is the sixth occurrence of that codeblock

File: RetailCoder.VBE/UI/RubberduckUI.de.resx [hunk 4]
```diff
@@ -1963,104 +1963,101 @@ Möchten sie die Einstellungen in Rubberduck importieren?</value>
...
+  <data name="ToolsMenu_ExportProject" xml:space="preserve">
     <value>Aktives Projekt exportieren</value>
   </data>
-  <data name="ImportCommand_OpenDialog_Filter_VBFiles">
+  <data name="ImportCommand_OpenDialog_Filter_VBFiles" xml:space="preserve">
     <value>VB Dateien</value>
   </data>
-  <data name="GeneralSettings_EnableSourceControl">
-    <value>Versionsverwaltung aktivieren. Benötigt Neustart zur Anwendung.</value>
-  </data>
```
Comment (Vogel612) [inline RetailCoder.VBE/UI/RubberduckUI.de.resx]: why was this dropped?

File: RetailCoder.VBE/UI/RubberduckUI.fr.resx [hunk 5]
```diff
@@ -1872,198 +1872,195 @@ End Sub</value>
...
+  <data name="ToolsMenu_ExportProject" xml:space="preserve">
     <value>Exporter le projet actif...</value>
   </data>
-  <data name="ImportCommand_OpenDialog_Filter_VBFiles">
+  <data name="ImportCommand_OpenDialog_Filter_VBFiles" xml:space="preserve">
     <value>Fichiers VB</value>
   </data>
-  <data name="GeneralSettings_EnableSourceControl">
-    <value>Activer le contrôle des source. Prendra effet au prochain démarrage.</value>
-  </data>
```
Comment (Vogel612) [inline RetailCoder.VBE/UI/RubberduckUI.fr.resx]: Why was this dropped?

Comment (retailcoder) [reply RetailCoder.VBE/UI/RubberduckUI.fr.resx]: @Vogel612 these resource keya are basically stale now that individual features can be toggled. Also there should be a dedicated settings tab for this, so the key would become `ExperimentalFeatureSettings_SourceControl` or something like it.

File: RetailCoder.VBE/Root/RubberduckIoCInstaller.cs (hunk 2 above)
Comment (Hosch250) [reply RetailCoder.VBE/Root/RubberduckIoCInstaller.cs]: Probably.

File: RetailCoder.VBE/UI/RubberduckUI.de.resx (hunk 4 above)
Comment (Hosch250) [reply RetailCoder.VBE/UI/RubberduckUI.de.resx]: Dropped in initial PR, then readded. I'll reintroduce it.

**Pull request merged**